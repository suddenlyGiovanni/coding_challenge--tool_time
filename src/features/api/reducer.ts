import { Reducer } from 'redux'
import { ActionType } from 'typesafe-actions'
import * as actions from './actions'

export type ApiAction = ActionType<typeof actions>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ApiState = any[]

const initialState: ApiState = []

export const apiReducer: Reducer<ApiState, ApiAction> = (
  state = initialState,
  action
) => {
  if (action.type.includes('@api/')) {
    return [...state, action]
  }
  return state
}
