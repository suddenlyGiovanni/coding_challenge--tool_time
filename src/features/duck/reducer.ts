import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'
import * as duckActions from './actions'

/**
 * @description Reducers
 * It's a good practice to keep your **state shape** in a comment above the reducers, just to have
 * an overview.
 *
 * In case the state shape is more complex, you should break the reducers into multiple smaller
 * functions that deal with a slice of the state, then combine them at the end.
 *
 * NOTE: Let's keep it simple for now with `switch` statements and abstract later.
 */

export type DucksAction = ActionType<typeof duckActions>

export type DucksState = Readonly<{
  quacking: boolean
  distance: number
  response?: string
}>

const initialState: DucksState = {
  quacking: false,
  distance: 0,
}

export const duckReducer: Reducer<DucksState, DucksAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case getType(duckActions.quack): {
      return { ...state, quacking: !state.quacking }
    }

    case getType(duckActions.swim): {
      const { distance } = action.payload
      return { ...state, distance: state.distance + distance }
    }

    case getType(duckActions.setDucks): {
      return { ...state, ...action.payload }
    }

    default: {
      return state
    }
  }
}

// export const duckReducer = createReducer<DucksState>(initialState, {
//   [getType(actions.quack)]: (state, action) => {
//     state.quacking = !state.quacking
//   },
//   [getType(actions.swim)]: (state, action) => {
//     const { distance } = action.payload
//     state.distance = state.distance + distance
//   },
// })
