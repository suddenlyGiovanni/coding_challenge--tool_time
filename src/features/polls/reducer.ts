import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'
import * as pollsActions from './actions'
import * as PollsTypes from './typings'

/**
 * @description Reducers
 * It's a good practice to keep your **state shape** in a comment above the reducers, just to have
 * an overview.
 *
 * In case the state shape is more complex, you should break the reducers into multiple smaller
 * functions that deal with a slice of the state, then combine them at the end.
 *
 * NOTE: Let's keep it simple for now with `switch` statements and abstract later.
 */

export type PollsAction = ActionType<typeof pollsActions>

export type PollsState = Readonly<{
  questions: PollsTypes.Questions
  votes: PollsTypes.Choice[]
}>

const initialState: PollsState = {
  questions: [],
  votes: [],
}

export const pollsReducer: Reducer<PollsState, PollsAction> = (
  state = initialState,
  action
) => {
  if (action.type === getType(pollsActions.setQuestions)) {
    return {
      ...state,
      questions: [...action.payload.questions],
    }
  }

  if (action.type === getType(pollsActions.setVoteResult)) {
    const { choice } = action.payload
    return {
      ...state,
      votes: [...state.votes, choice],
    }
  }
  return state
}
