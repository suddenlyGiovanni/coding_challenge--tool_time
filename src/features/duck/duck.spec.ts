import {
  duckActions as actions,
  duckReducer as reducer,
  duckTypes as types,
  duckSelectors as selectors,
} from '.'
import { DucksState } from './reducer'
import { RootState } from 'typesafe-actions'

/**
 * FIXTURES
 */
const getInitialState = (initial?: Partial<DucksState>) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return reducer(initial as DucksState, {} as any)
}

const getRootState = (initial?: Partial<DucksState>): RootState => ({
  duck: getInitialState(initial),
})

describe('feature: `duck`', () => {
  describe('actions', () => {
    it('should create an action to make a duck quack', () => {
      const expectedAction = { type: types.QUACK }
      expect(actions.quack()).toEqual(expectedAction)
    })

    it('should create an action to make a duck swim', () => {
      const distance = 100
      const expectedAction = { type: types.SWIM, payload: { distance } }
      expect(actions.swim(distance)).toEqual(expectedAction)
    })
  })

  describe('reducer', () => {
    describe('initial state', () => {
      it('should match a snapshot', () => {
        const initialState = getInitialState()
        expect(initialState).toMatchSnapshot()
      })
    })

    describe('toggling `quack`', () => {
      const initialState = getInitialState()
      const state1 = reducer(initialState, actions.quack())
      const state2 = reducer(state1, actions.quack())

      test('the initial state of quacking should be `false`', () => {
        expect(initialState.quacking).toBe(false)
      })

      it('should toggle `quacking` from `false` to `true`', () => {
        expect(state1.quacking).toBe(true)
      })

      it('should toggle `quacking` from `true` to `false`', () => {
        expect(state2.quacking).toBe(false)
      })
    })

    describe('swimming', () => {
      it('should `sum` the `initial` distance `to` the `new` swam `distance`', () => {
        const distance = 100
        const initialState = getInitialState()
        const state1 = reducer(initialState, actions.swim(distance))
        const state2 = reducer(state1, actions.swim(distance))
        expect(initialState.distance).toBe(0)
        expect(state1.distance).toBe(distance)
        expect(state2.distance).toBe(distance + distance)
      })
    })
  })

  describe('selectors', () => {
    const initialRootState = getRootState()

    it('checkIfDuckIsInRange', () => {
      const state1 = getInitialState({ distance: 1001 }) //?
      const rootState1 = getRootState(state1)

      expect(selectors.checkIfDuckIsInRange(initialRootState)).toBe(false)
      expect(selectors.checkIfDuckIsInRange(rootState1)).toBe(true)
    })

    it('checkIfDuckIsQuaking', () => {
      expect(selectors.checkIfDuckIsQuaking(initialRootState)).toBe(false)
      const state1 = getInitialState({ quacking: true })
      const rootState1 = getRootState(state1)
      expect(selectors.checkIfDuckIsQuaking(rootState1)).toBe(true)
    })

    it('duckDistance', () => {
      const distance = 100
      expect(selectors.duckDistance(initialRootState)).toBe(0)
      expect(
        selectors.duckDistance(getRootState(getInitialState({ distance })))
      ).toBe(distance)
    })
  })
})
