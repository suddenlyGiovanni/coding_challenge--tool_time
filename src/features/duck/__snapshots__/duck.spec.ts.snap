// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature: \`duck\` reducer initial state should match a snapshot 1`] = `
Object {
  "distance": 0,
  "quacking": false,
}
`;
